### DISCUSSION TITLE
Project 3 - confused by instructions

### DISCUSSION THREAD

**Discussion by 2199105T** [IBAHCM/RPiR#341] [Archive] [answered]
Hi @richardreeve @soniamitchell,

There is a little confusion in the instructions for project 3 that I was hoping you would be able to clarify. The text in the exercise is:

```{r}
Then you need to write two new very simple functions which calculate the Shannon entropy and the Berger-Parker index, using the generic function you created above and not by calculating them from scratch (they will also need the small amount of additional calculation – Shannon entropy is actually log(𝐷1) and Berger-Parker is 1𝐷∞ – as detailed above in the overview).
```

Where it says 'they will also need a small amount of additional calculation...' I am unsure where this calculation should be. To yield the same result it doesn't make sense to have it in both the specific and general functions, as you would be doing the calculation twice when testing the specific function.

So from the instructions, it isn't clear whether this additional calculation should be in the general function or the specific functions, are you able to clarify which it should be? or does it not really matter where this additional calculation is as long as you yield the same  result for both general and specific.

**Comment by richardreeve** [comment, marked as answer]
So, your general function should always calculate `D(q)` for whatever value of `q` it is given - that's just what it does. However the Shannon entropy function must calculate Shannon entropy (of course) and the two are not the same. So your Shannon entropy has to do a bit more work (the "additional calculation") after it has received `D(1)` from your diversity function. Does that make sense?

**Reply by 2199105T** [reply]
Hi Richard,

I think I need some more clarification. If the additional calculations are needed in both the general and the specific function, wouldn't this mean that we would get an incorrect result when calling the specific function that uses the general one? Wouldn't that do calculations / calculate log twice? I thought that in order to calculate D we need to calculate Shannon or Berger itself, so I am unsure how to calculate D in the general function without yielding incorrect results for the specific functions.

**Reply by richardreeve** [reply]
No, the additional instructions are never needed in the general functions. The general function is as we explained when we described the general function. Shannon, Berger-Parker, etc. are all slight adaptations of the general functions.
